keanu-project/src/test/java/io/improbable/keanu/algorithms/mcmc/MetropolisHastingsTest.java

```diff
@@ -95,63 +113,62 @@ public void samplesContinuousTensorPrior() {
+        MCMCTestCase testCase = new SingleVariateDiscreteTestCase();
 
-        BernoulliVertex C = new BernoulliVertex(B);
-
-        C.observe(true);
-
-        BayesianNetwork bayesNet = new BayesianNetwork(Arrays.asList(A, B, C));
-        bayesNet.probeForNonZeroProbability(100);
+        BayesianNetwork bayesNet = testCase.getModel();
         KeanuProbabilisticModel model = new KeanuProbabilisticModel(bayesNet);
 
         NetworkSamples posteriorSamples = Keanu.Sampling.MetropolisHastings.withDefaultConfigFor(model).getPosteriorSamples(
             model,
-            Collections.singletonList(A),
+            model.getLatentVertices(),
             10000
         );
 
-        double postProbTrue = posteriorSamples.get(A).probability(v -> v.scalar());
-
-        assertEquals(0.9, postProbTrue, 0.01);
+        testCase.assertExpected(posteriorSamples);
     }
 
     @Category(Slow.class)
     @Test
-    public void samplesComplexDiscretePrior() {
-
-        BernoulliVertex A = new BernoulliVertex(0.5);
-        BernoulliVertex B = new BernoulliVertex(0.5);
-
-        BooleanVertex C = A.or(B);
+    public void samplesComplexDiscretePriorWithDefaults() {
 
-        DoubleVertex D = If.isTrue(C)
-            .then(0.9)
-            .orElse(0.1);
+        MCMCTestCase testCase = new MultiVariateDiscreteTestCase();
 
-        BernoulliVertex E = new BernoulliVertex(D);
-
-        E.observe(true);
-
-        BayesianNetwork bayesNet = new BayesianNetwork(Arrays.asList(A, B, C, D, E));
-        bayesNet.probeForNonZeroProbability(100);
+        BayesianNetwork bayesNet = testCase.getModel();
         KeanuProbabilisticModel model = new KeanuProbabilisticModel(bayesNet);
 
         NetworkSamples posteriorSamples = Keanu.Sampling.MetropolisHastings.withDefaultConfigFor(model).getPosteriorSamples(
             model,
-            Collections.singletonList(A),
+            model.getLatentVertices(),
             1000
         );
 
-        double postProbTrue = posteriorSamples.get(A).probability(v -> v.scalar());
+        testCase.assertExpected(posteriorSamples);
+    }
+
+    @Category(Slow.class)
+    @Test
+    public void samplesComplexDiscreteWithFullVariableSelect() {
+
+        MCMCTestCase testCase = new MultiVariateDiscreteTestCase();
 
-        assertEquals(0.643, postProbTrue, 0.01);
+        BayesianNetwork bayesNet = testCase.getModel();
+
+        KeanuProbabilisticModel model = new KeanuProbabilisticModel(bayesNet);
+
+        NetworkSamples posteriorSamples = MetropolisHastings.builder()
+            .proposalDistribution(new PriorProposalDistribution(model.getLatentVertices()))
+            .rejectionStrategy(new RollBackToCachedValuesOnRejection(model.getLatentVertices()))
+            .variableSelector(MHStepVariableSelector.FULL_VARIABLE_SELECTOR)
+            .build()
+            .getPosteriorSamples(
+                model,
+                model.getLatentVertices(),
+                1000
+            );
```

Comment:
There's quite a lot of fixed structure to a lot of these test cases now - could we turn it in to a single function that each test case calls?